File: agent/src/main/java/com/asura/agent/util/SocketSendUtil.java
package com.asura.agent.util;

import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;
import com.asura.agent.conf.MonitorCacheConfig;
import com.asura.agent.controller.MonitorController;
import com.asura.agent.entity.PushServerEntity;
import com.asura.agent.thread.SocketThread;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.lang.reflect.Type;
import java.net.InetAddress;
import java.util.ArrayList;
import java.util.List;
import java.util.Random;

/**
 * <p></p>
 *
 * <PRE>
 * <BR>	修改记录
 * <BR>-----------------------------------------------
 * <BR>	修改日期			修改人			修改内容
 * </PRE>
 *
 * @author zhaozq
 * @version 1.0
 * @since 1.0
 */

public class SocketSendUtil {

    public static final Logger logger = LoggerFactory.getLogger(SocketSendUtil.class);
    private static InetAddress server;
    private static ArrayList<InetAddress> serverList;
    private static int port;
    private static ArrayList<Integer> ports;
    private static RedisUtil redisUtil;
    private static Random random;

    /**
     * 获取服务器列表大小
     *
     * @return
     */
    public static int getServerListSize() {
        if (serverList != null) {
            return serverList.size();
        } else {
            return 0;
        }
    }

    /**
     * 先优先使用redis的
     */
    static void setServerList() {
        InetAddress addr;
        if (redisUtil == null) {
            redisUtil = new RedisUtil();
        }
        if (serverList == null ){
            serverList = new ArrayList<>();
        }
        String pushServers = redisUtil.get(MonitorCacheConfig.cachePushServer);
        if (pushServers != null && pushServers.length() > 0) {
            Type type = new TypeToken<ArrayList<PushServerEntity>>() {
            }.getType();
            List<PushServerEntity> serverEntities = new Gson().fromJson(pushServers, type);
            for (PushServerEntity serverEntity : serverEntities) {
                try {
                    addr = InetAddress.getByName(serverEntity.getIp());
                    if (!serverList.contains(addr)) {
                        logger.info("添加push服务Redis " + addr);
                        serverList.add(addr);
                    }
                } catch (Exception e) {
                    e.printStackTrace();
                    logger.error("获取push server失败 1", e);
                }
            }
        }
    }

    /**
     * 随机链接服务器
     *
     * @return
     */
    public static InetAddress getServer(InetAddress address) {
        if (MonitorController.getErrorNumber()){
            return null;
        }
        if (serverList != null && serverList.size() > 0) {
            if (address == null) {
                random = new Random();
                int id = random.nextInt(serverList.size());
                return serverList.get(id);
            }
            // 去除坏了的连接
            for (int i = 0; i < serverList.size(); i++) {
                if (serverList.get(i) != address) {
                    return serverList.get(i);
                }
            }
        } else {
            serverList = new ArrayList<>();
            setServerList();
            return getServer(null);
        }
        return null;
    }

    /**
     * @param addr
     *
     * @return
     */
    static boolean checkServerValid(String addr) {
        String port = redisUtil.get(MonitorCacheConfig.getCachePushServerPort.concat(addr.toString().replace("/", "")).trim());
        if (port.length() > 1) {
            return true;
        }
        return false;
    }

    /**
     * 每10分钟重新获取一次
     * 设置上传服务器
     */
    public static void setServer() {
        serverList = new ArrayList<>();
        RedisUtil redisUtil = new RedisUtil();
        String key = MonitorCacheConfig.cachePushServer;
        String addrs = redisUtil.get(key);
        if (addrs != null && addrs.length() > 8) {
            Type type = new TypeToken<ArrayList<PushServerEntity>>() {
            }.getType();
            List<PushServerEntity> list = new Gson().fromJson(addrs, type);
            for (PushServerEntity entity : list) {
                try {
                    InetAddress address = InetAddress.getByName(entity.getIp());
                    if (!serverList.contains(address)) {
                        if (SocketThread.sendData("[{}]", address, 50329) && checkServerValid(address.toString())) {
                            logger.info("获取到PUSH服务器地址" + entity.getIp());
                            serverList.add(address);
                        }
                    }
                } catch (Exception e) {
                }
            }
        }
    }

    public static void init() {
        if (redisUtil == null) {
            redisUtil = new RedisUtil();
        }
        if (serverList == null) {
            setServerList();
        }
        if (random == null) {
            random = new Random();
        }
        try {
            if (ports == null) {
                ports = new ArrayList();
                for (int i = 50000; i < 50300; i++) {
                    ports.add(i);
                }
            }
        } catch (Exception e) {
            logger.info("ERROR:获取服务地址失败" + "server");
        }
    }

    /**
     * @param data
     *
     * @return
     */
    public static void sendData(String data) {
        init();
        server = getServer(null);
        if (server == null) {
            logger.error("获取push服务失败..");
            return;
        }
        port = ports.get(random.nextInt(299));
        SocketThread thread = new SocketThread(data, server, port);
        thread.start();
        long start = System.currentTimeMillis() / 1000;
        while (1 == 1) {
            if (!thread.isAlive()) {
                break;
            }
            if (System.currentTimeMillis() / 1000 - start > 2) {
                logger.error("udp线程超时");
                if (thread.isAlive()) {
                    try {
                        thread.interrupt();
                    } catch (Exception e) {
                    }
                }
                break;
            }
            try {
                Thread.sleep(10);
            } catch (Exception e) {
            }
        }
    }

}

File: monitor-agent-go/src/service/socketSendUtil.go
package service

import (
	"util/redisUtil"
	"configure"
	"encoding/json"
	"math/rand"
	"strconv"
	"strings"
	"util/commonUtil"
)

var (
	serverList []string
	ports      []string
	Url        string
)

type PushServerEntity struct {
	Ip string
}

func Init() {
	if len(serverList) < 1 {
		setServerList()
	}
	//num := rand.Intn(len(serverList))
	if len(ports) < 1 {
		ports = make([]string, 0)
		for i := 50000; i < 50300; i++ {
			ports = append(ports, strconv.Itoa(i))
		}
	}
}

/**
   * 先优先使用redis的
   */
func setServerList() {
	if len(serverList) < 1 {
		serverList = make([]string, 0)
	}
	pushServers := redisUtil.Get_del("get", configure.CachePushServer)
	if commonUtil.CheckString(pushServers) {
		var serverEntities []PushServerEntity
		json.Unmarshal([]byte(pushServers), &serverEntities)
		for _, serverEntity := range serverEntities {
			addr := serverEntity.Ip
			for _, ip := range serverList {
				if ip == addr {
					continue
				}
			}
			Info("添加push服务Redis " + addr)
			serverList = append(serverList, addr)
		}
	}
}

/**
 * 随机链接服务器
 *
 * @return
 */
func GetServer(address string) string {
	if GetErrorNumber() {
		Info("获取到redis失败，程序不再执行")
		return ""
	}
	if len(serverList) > 0 {
		if address == "" {
			id := rand.Intn(len(serverList))
			return serverList[id]
		}
		// 去除坏了的连接
		for i := 0; i < len(serverList); i++ {
			if serverList[i] != address {
				return serverList[i]
			}
		}
	} else {
		serverList = make([]string, 0)
		setServerList()
		return ""
	}
	return ""
}

/**
 * 每10分钟重新获取一次
 * 设置上传服务器
 */
func SetServer() {
	serverList = make([]string, 0)
	key := configure.CachePushServer
	addrs := redisUtil.Get_del("get", key)
	if len(addrs) > 8 {
		var list = []PushServerEntity{}
		json.Unmarshal([]byte(addrs), &list)
		for _, entity := range list {
			address := entity.Ip
			isExists := false
			for _, add := range serverList {
				if add == address {
					isExists = true
				}
			}
			if isExists {
				continue
			}
			if SendData("[{}]", address, strconv.Itoa(50329)) && checkServerValid(address) {
				Info("获取到PUSH服务器地址" + entity.Ip)
				serverList = append(serverList, address)
			}
		}
	}
}

func checkServerValid(addr string) bool {
	port := redisUtil.Get_del("get", configure.GetCachePushServerPort+addr)
	port = strings.Replace(port, "/", "", -1)
	port = strings.Replace(port, " ", "", -1)
	if commonUtil.CheckString(port) {
		return true
	}
	return false
}

/**
 * 初始化server
 */
func SetHttpServer(failServer string) {
	address := GetServer(failServer)
	if !commonUtil.CheckString(address) {
		return
	}
	//Server := address
	port := redisUtil.Get_del("get", configure.GetCachePushServerPort+address)
	Url = "http://" + address + ":" + port + "/"
	Info(Url)
}

func getServerListSize() int {
	return len(serverList)
}
